### PR TITLE
feat: configure max block range for eth_getLogs

### PR DESCRIPTION
**Pull request by NazariiDenha**
## 1. Purpose or design rationale of this PR

limit `eth_getLogs` block range
https://app.asana.com/0/1202293017617135/1205536550847251/f

## 2. PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [ ] build: Changes that affect the build system or external dependencies (example scopes: yarn, eslint, typescript)
- [ ] ci: Changes to our CI configuration files and scripts (example scopes: vercel, github, cypress)
- [ ] docs: Documentation-only changes
- [x] feat: A new feature
- [ ] fix: A bug fix
- [ ] perf: A code change that improves performance
- [ ] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance
- [ ] style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [ ] test: Adding missing tests or correcting existing tests

## 3. Deployment tag versioning

Has the version in `params/version.go` been updated?

- [ ] This PR doesn't involve a new deployment, git tag, docker image tag, and it doesn't affect traces
- [x] Yes

## 4. Breaking change label

Does this PR have the `breaking-change` label?

- [x] This PR is not a breaking change
- [ ] Yes

### CODE REVIEW

File: eth/filters/api.go [hunk 1]
```diff
@@ -345,6 +347,9 @@ func (api *PublicFilterAPI) GetLogs(ctx context.Context, crit FilterCriteria) ([
 		if crit.ToBlock != nil {
 			end = crit.ToBlock.Int64()
 		}
+		if end-begin+1 > api.maxBlockRange {
```
Comment (colinlyguo) [inline eth/filters/api.go]: can add a unit test here. e.g., `api.maxBlockRange` is 2, test fetch 2 (happy path) and fetch 3 (error path).

File: eth/backend.go [hunk 2]
```diff
@@ -333,7 +333,7 @@ func (s *Ethereum) APIs() []rpc.API {
 		}, {
 			Namespace: "eth",
 			Version:   "1.0",
-			Service:   filters.NewPublicFilterAPI(s.APIBackend, false, 5*time.Minute),
+			Service:   filters.NewPublicFilterAPI(s.APIBackend, false, 5*time.Minute, s.config.MaxBlockRange),
```
Comment (colinlyguo) [inline eth/backend.go:336]: Would this value be `0`?

File: eth/filters/api.go (hunk 1 above)
Comment (Thegaram) [inline eth/filters/api.go]: This won't work if `begin` or `end` use `rpc.LatestBlockNumber`, `rpc.EarliestBlockNumber`, etc.

Comment (NazariiDenha) [reply eth/filters/api.go]: Why? Do we want to check that requested range is not bigger or real fetched?

Comment (Thegaram) [reply eth/filters/api.go]: If `begin == rpc.EarliestBlockNumber` and `end == rpc.LatestBlockNumber`, then the actual range can be huge but this check will pass.

Comment (NazariiDenha) [reply eth/filters/api.go]: added a test

File: eth/backend.go (hunk 2 above)
Comment (NazariiDenha) [reply eth/backend.go:336]: added a default value

File: eth/filters/api.go (hunk 1 above)
Comment (NazariiDenha) [reply eth/filters/api.go]: updated

**Review by colinlyguo — DISMISSED** [review]

File: eth/filters/api.go [hunk 3]
```diff
@@ -345,6 +347,20 @@ func (api *PublicFilterAPI) GetLogs(ctx context.Context, crit FilterCriteria) ([
...
+		}
+		endBlock, err := api.backend.HeaderByNumber(ctx, rpc.BlockNumber(end))
+		if err != nil {
+			return nil, fmt.Errorf("couldn't find toBlock, toBlock: %d", end)
+		}
+		realBegin := beginBlock.Number.Int64()
+		realEnd := endBlock.Number.Int64()
+		if realEnd-realBegin+1 > api.maxBlockRange {
+			return nil, fmt.Errorf("block range is bigger than maxBlockRange, block range: %d, maxBlockRange: %d", realEnd-realBegin+1, api.maxBlockRange)
+		}
```
Comment (georgehao) [inline eth/filters/api.go:363]: If the maxBlockRange reached, why not return the begin to maxBlockRang instead of returning the error

Comment (Thegaram) [reply eth/filters/api.go:363]: That is very confusing behaviour, the caller has no way to know that they did not get all the data that they asked for.

Comment (georgehao) [reply eth/filters/api.go:363]: ok, get it.

**Review by georgehao — DISMISSED** [review]

File: eth/filters/filter_system_test.go [hunk 4]
```diff
@@ -168,7 +169,7 @@ func TestBlockSubscription(t *testing.T) {
 	var (
 		db          = rawdb.NewMemoryDatabase()
 		backend     = &testBackend{db: db}
-		api         = NewPublicFilterAPI(backend, false, deadline)
+		api         = NewPublicFilterAPI(backend, false, deadline, 5000)
```
Comment (mask-pp) [inline eth/filters/filter_system_test.go]:
```suggestion
		api         = NewPublicFilterAPI(backend, false, deadline, Defaults.MaxBlockRange)
```

File: eth/filters/filter_system_test.go [hunk 5]
```diff
@@ -220,7 +221,7 @@ func TestPendingTxFilter(t *testing.T) {
 	var (
 		db      = rawdb.NewMemoryDatabase()
 		backend = &testBackend{db: db}
-		api     = NewPublicFilterAPI(backend, false, deadline)
+		api     = NewPublicFilterAPI(backend, false, deadline, 5000)
```
Comment (mask-pp) [inline eth/filters/filter_system_test.go]: ditto

**Review by mask-pp — DISMISSED** [review]
LGTM just some carps.

File: eth/ethconfig/config.go [hunk 6]
```diff
@@ -210,6 +211,9 @@ type Config struct {
 
 	// Check circuit capacity in block validator
 	CheckCircuitCapacity bool
+
+	// Max block range for eth_getLogs api method
+	MaxBlockRange int64
```
Comment (Thegaram) [inline eth/ethconfig/config.go:216]: Add a CLI param, e.g. `--rpc.getlogs.maxrange` (see how `CheckCircuitCapacity is set).

**Comment by colinlyguo** [comment]
CI failed with: `Error: cmd/utils/flags.go:1538:6: `setMaxBlockRange` is unused (deadcode)`

File: cmd/utils/flags.go [hunk 7]
```diff
@@ -831,6 +831,12 @@ var (
 		Name:  "ccc",
 		Usage: "Enable circuit capacity check during block validation",
 	}
+
+	// Max block range for `eth_getLogs` method
+	MaxBlockRangeFlag = cli.Int64Flag{
+		Name:  "rpc.getlogs.maxrange",
+		Usage: "Limit max fetched block range for `eth_GetLogs` method",
```
Comment (colinlyguo) [inline cmd/utils/flags.go]:
```suggestion
		Usage: "Limit max fetched block range for `eth_getLogs` method",
```

File: eth/ethconfig/gen_config.go [hunk 8]
```diff
@@ -60,6 +60,9 @@ func (c Config) MarshalTOML() (interface{}, error) {
 		Checkpoint              *params.TrustedCheckpoint      `toml:",omitempty"`
 		CheckpointOracle        *params.CheckpointOracleConfig `toml:",omitempty"`
 		OverrideArrowGlacier    *big.Int                       `toml:",omitempty"`
+		MPTWitness              int
```
Comment (colinlyguo) [inline eth/ethconfig/gen_config.go:63]: What are `MPTWitness` and `CheckCircuitCapacity` used for here? Add missed items in marshaling/unmarshaling config files?

**Review by colinlyguo — DISMISSED** [review]
LGTM

**Comment by colinlyguo** [comment]
Great work! I noticed that this change is implemented directly at the node level instead of in the `rpc-gateway`. While both approaches seem viable, I am keen to understand the rationale behind choosing to modify it at the node level rather than at the rpc-gateway.

File: eth/ethconfig/gen_config.go (hunk 8 above)
Comment (NazariiDenha) [reply eth/ethconfig/gen_config.go:63]: yes

**Comment by NazariiDenha** [comment]
> Great work! I noticed that this change is implemented directly at the node level instead of in the `rpc-gateway`. While both approaches seem viable, I am keen to understand the rationale behind choosing to modify it at the node level rather than at the rpc-gateway.

I think, at least it's harder and less efficient to get real block number (if arg is one of `finalized`, `safe` etc.) from `rpc-gateway`)

**Comment by colinlyguo** [comment]
> > Great work! I noticed that this change is implemented directly at the node level instead of in the `rpc-gateway`. While both approaches seem viable, I am keen to understand the rationale behind choosing to modify it at the node level rather than at the rpc-gateway.
> 
> I think, at least it's harder and less efficient to get real block number (if arg is one of `finalized`, `safe` etc.) from `rpc-gateway`)

Yeah. Sounds reasonable.

**Review by colinlyguo — DISMISSED** [review]

**Review by colinlyguo — APPROVED** [review]

**Review by georgehao — APPROVED** [review]

**Pull request merged**